File: cpp_solutions/1219.cpp
class Solution {
public:
    vector<pair<int, int>> directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

    int dfs(vector<vector<int>>& grid, int i, int j) {
        int m = grid.size();
        int n = grid[0].size();
        int current = grid[i][j];
        int maxValue = current;
        grid[i][j] = 0;
        for (auto direction : directions) {
            int x = i + direction.first;
            int y = j + direction.second;
            if (x < 0 || x >= m || y < 0 || y >= n) continue;
            if (grid[x][y] == 0) continue;
            maxValue = max(maxValue, current + dfs(grid, x, y));
        }
        grid[i][j] = current;
        return maxValue;
    }
    int getMaximumGold(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        int maxValue = 0;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == 0) continue;
                cout << "Now " << i << " " << " " << j << " " << maxValue << endl;
                maxValue = max(maxValue, dfs(grid, i, j));
            }
        }
        return maxValue;
    }
};

File: rust_solutions/1219.rs
impl Solution {
    pub fn get_maximum_gold(grid: Vec<Vec<i32>>) -> i32 {
        let (m, n) = (grid.len(), grid[0].len());

        let mut grid_copy = grid.clone();
        let mut_grid: &mut Vec<Vec<i32>> = &mut grid_copy;
        fn dfs(grid: &mut Vec<Vec<i32>>, r: usize, c: usize) -> i32 {
            let (m, n) = (grid.len(), grid[0].len());
            let current = grid[r][c];
            let mut max_value = current;
            grid[r][c] = 0;
            for (x, y) in [(r+1, c), (r-1, c), (r, c+1), (r, c-1)] {
                if x < 0 || x >= m || y < 0 || y >= n {
                    continue;
                }
                let x = x as usize;
                let y = y as usize;
                if grid[x][y] == 0 {
                    continue;
                }
                max_value = max_value.max(current + dfs(grid, x, y));
            }
            grid[r][c] = current;
            max_value
        }

        let mut res = 0;
        for r in 0..m {
            for c in 0..n {
                if grid[r][c] != 0 {
                    res = res.max(dfs(mut_grid, r, c));
                }
            }
        }
        res
    }
}
